File: MyHome_Android/home/src/main/java/com/my/home/Utils/XmlSerializer.java
package com.my.home.Utils;

import android.content.Context;
import android.graphics.Point;

import com.my.home.TcpConnection.Client;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.lang.reflect.ParameterizedType;
import java.util.Arrays;
import java.util.List;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.Result;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

/**
 * Created by Hristo on 28.6.2014 г..
 */
public class XmlSerializer {

    @Target({ElementType.FIELD, ElementType.TYPE})
    @Retention(RetentionPolicy.RUNTIME)
    public static @interface XmlIgnoreAttribute {
    }

    @Target(ElementType.FIELD)
    @Retention(RetentionPolicy.RUNTIME)
    public static @interface XmlIdentifierAttribute {
    }


    public static Document newDocument() {
        Document xmlDoc = null;
        try {
            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            xmlDoc = builder.newDocument();
        } catch (Exception e) {
			Client.log("Exception: " + e.getMessage()); 
        }
        return xmlDoc;
    }

    public static Document parseDocument(Context context, String fileName) {
        Document xmlDoc = null;
        try {
            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            xmlDoc = builder.parse(context.openFileInput(fileName));
        } catch (Exception e) {
            Client.log("Exception: " + e.getMessage());
        }
        return xmlDoc;
    }

    public static void saveDocument(Context context, String fileName, Document xmlDoc) {
        try {
            Transformer transformer = TransformerFactory.newInstance().newTransformer();
            Result result = new StreamResult(context.openFileOutput(fileName, Context.MODE_PRIVATE));
            transformer.transform(new DOMSource(xmlDoc), result);
        } catch (Exception e) {
            Client.log("Exception: " + e.getMessage());
        }
    }


    public static void serialize(Document xmlDoc, Object obj) throws IllegalAccessException {
        Element xmlRoot = xmlDoc.getDocumentElement();
        XmlSerializer.serialize(xmlRoot, obj);
    }

    public static void serialize(Element xmlRoot, Object obj) throws IllegalAccessException {
        Class type = obj.getClass();

        Document xmlDoc = xmlRoot.getOwnerDocument();
        Element xmlMain = xmlDoc.createElement(type.getSimpleName());
        xmlRoot.appendChild(xmlMain);

        if (type.getAnnotation(XmlIgnoreAttribute.class) != null)
            return;

        if (type.equals(String.class)) {
            xmlMain.setAttribute("value", obj.toString());
            return;
        }

        Field[] fields = type.getDeclaredFields();
        for (Field field : fields) {
            if (Modifier.isStatic(field.getModifiers()))
                continue;
             if (field.getAnnotation(XmlIgnoreAttribute.class) != null)
                continue;

            field.setAccessible(true);
			if (field.getType().isArray()) {
                Object array = field.get(obj);
                String value = "";
                for (int i = 0; i < Array.getLength(array); i++)
                    value += Array.get(array, i).toString() + ",";
                if (value.length() > 0)
                    value = value.substring(0, value.length() - 1);
                xmlMain.setAttribute(field.getName(), value);
            }
            else if (List.class.isAssignableFrom(field.getType()) && !field.getType().equals(String.class)) {
                Element xmlElement = xmlDoc.createElement(field.getName());
                xmlMain.appendChild(xmlElement);

                List list = (List)field.get(obj);
                for (Object elem : list)
                    XmlSerializer.serialize(xmlElement, elem);
            }
            else if (field.getType().equals(int.class) && field.getAnnotation(XmlIdentifierAttribute.class) != null) {
                int value = Math.abs((Integer) field.get(obj));
                xmlMain.setAttribute(field.getName(), String.format("%010d", value));
            }
            else if (field.getType().equals(Point.class)) {
                Point value = (Point) field.get(obj);
                xmlMain.setAttribute(field.getName(), value.x + "," + value.y);
            }
            else {
                Object value = field.get(obj);
                if (value != null)
                    xmlMain.setAttribute(field.getName(), value.toString());
            }
        }
    }


    public static void deserialize(Document xmlDoc, Object obj) throws IllegalAccessException {
        Element xmlRoot = xmlDoc.getDocumentElement();
        if (xmlRoot == null) return;
        Element xmlMain = (Element)xmlRoot.getElementsByTagName(obj.getClass().getSimpleName()).item(0);
        if (xmlMain == null) return;

        XmlSerializer.deserialize(xmlMain, obj);
    }

    public static void deserialize(Element xmlMain, Object obj) throws IllegalAccessException {
        Class type = obj.getClass();

        NamedNodeMap attributes = xmlMain.getAttributes();
        for (int i = 0; i < attributes.getLength(); i++) {
            Node xmlAttribute = attributes.item(i);
            Field field = null;
            try {
                field = type.getDeclaredField(xmlAttribute.getNodeName());
            } catch (Exception e) { 
				Client.log("Exception: " + e.getMessage()); 
			}
            if (field == null)
                continue;

            field.setAccessible(true);
			if (field.getType().isArray()) {
                String[] values = xmlAttribute.getNodeValue().split(",");
                Class elementType = field.getType().getComponentType();
                Object array = Array.newInstance(elementType, values.length);
                for (int j = 0; j < values.length; j++)
                    Array.set(array, j, Utils.parse(values[j], elementType));
                field.set(obj, array);
			}
			else {
                field.set(obj, Utils.parse(xmlAttribute.getNodeValue(), field.getType()));
			}
        }

        NodeList childNodes = xmlMain.getChildNodes();
        for (int i = 0; i < childNodes.getLength(); i++) {
            Element xmlElement = Utils.as(childNodes.item(i), Element.class);
            if (xmlElement == null)
                continue;

            Field field = null;
            try {
                field = type.getDeclaredField(xmlElement.getNodeName());
            } catch (Exception e) {
				Client.log("Exception: " + e.getMessage());
			}
            if (field == null || !List.class.isAssignableFrom(field.getType()))
                continue;

            field.setAccessible(true);
            List list = (List)field.get(obj);
            list.clear();

            ParameterizedType genType = Utils.as(field.getGenericType(), ParameterizedType.class);
            if (genType != null && genType.getActualTypeArguments().length > 0) {
                NodeList childNodes2 = xmlElement.getChildNodes();
                for (int j = 0; j < childNodes2.getLength(); j++) {
                    Element xmlElement2 = Utils.as(childNodes2.item(j), Element.class);
                    if (xmlElement2 == null)
                        continue;

                    if (((Class) genType.getActualTypeArguments()[0]).equals(String.class))
                        list.add(xmlElement2.getAttribute("value"));
                    else {
                        Object newObj = null;
                        try {
                            newObj = ((Class) genType.getActualTypeArguments()[0]).newInstance();
                        } catch (Exception e) {
                            Client.log("Exception: " + e.getMessage());
                        }
                        XmlSerializer.deserialize(xmlElement2, newObj);
                        list.add(newObj);
                    }
                }
            }
        }
    }

}

File: MyHome/Utils/XmlSerializer.cs
using System;
using System.Collections;
using System.Globalization;
using System.Reflection;
using System.Xml;

namespace MyHome.Utils
{
    [AttributeUsage(AttributeTargets.Property)]
    public class XmlIgnoreAttribute : Attribute
    {
    }

    [AttributeUsage(AttributeTargets.Property)]
    public class XmlIdentifierAttribute : Attribute
    {
    }


    public static class XmlSerializer
    {

        public static void Serialize(XmlDocument xmlDoc, Object obj)
        {
            XmlElement xmlRoot = xmlDoc.DocumentElement;
            XmlSerializer.Serialize(xmlRoot, obj);
        }

        public static void Serialize(XmlElement xmlRoot, Object obj)
        {
            Type type = obj.GetType();

            XmlDocument xmlDoc = xmlRoot.OwnerDocument;
            XmlElement xmlMain = xmlDoc.CreateElement(type.Name);
            xmlRoot.AppendChild(xmlMain);

            if (type == typeof(string))
            {
                xmlMain.SetAttribute("Value", obj.ToString());
                return;
            }

            PropertyInfo[] pis = type.GetProperties();
            foreach (PropertyInfo pi in pis)
            {
                if (pi.GetCustomAttribute(typeof(XmlIgnoreAttribute)) != null)
                    continue;

                if (pi.PropertyType.IsArray)
                {
                    Array array = pi.GetValue(obj) as Array;
                    string value = "";
                    foreach (Object elem in array)
                        value += Utils.toInvariantString(elem) + ",";
                    if (value.Length > 0)
                        value = value.Substring(0, value.Length - 1);
                    xmlMain.SetAttribute(pi.Name, value);
                }
                else if (typeof(IList).IsAssignableFrom(pi.PropertyType) && pi.PropertyType != typeof(string))
                {
                    XmlElement xmlElement = xmlDoc.CreateElement(pi.Name);
                    xmlMain.AppendChild(xmlElement);

                    IList list = pi.GetValue(obj) as IList;
                    foreach (Object elem in list)
                        XmlSerializer.Serialize(xmlElement, elem);
                }
                else if (pi.PropertyType == typeof(int) && pi.GetCustomAttribute(typeof(XmlIdentifierAttribute)) != null)
                {
                    int value = Math.Abs((int)pi.GetValue(obj));
                    xmlMain.SetAttribute(pi.Name, value.ToString("D10", CultureInfo.InvariantCulture));
                }
                else if (pi.PropertyType == typeof(System.Drawing.Color))
                {
                    System.Drawing.Color value = (System.Drawing.Color)pi.GetValue(obj);
                    xmlMain.SetAttribute(pi.Name, Utils.toInvariantString(value));
                }
                else
                {
                    Object value = pi.GetValue(obj);
                    if (value != null)
                        xmlMain.SetAttribute(pi.Name, Utils.toInvariantString(value));
                }
            }
        }


        public static void Deserialize(XmlDocument xmlDoc, Object obj)
        {
            XmlElement xmlRoot = xmlDoc.DocumentElement;
            if (xmlRoot == null) return;
            XmlElement xmlMain = xmlRoot.GetElementsByTagName(obj.GetType().Name)[0] as XmlElement;
            if (xmlMain == null) return;

            XmlSerializer.Deserialize(xmlMain, obj);
        }

        public static void Deserialize(XmlElement xmlMain, Object obj)
        {
            Type type = obj.GetType();

            foreach (XmlAttribute xmlAttribute in xmlMain.Attributes)
            {
                PropertyInfo pi = type.GetProperty(xmlAttribute.Name);
                if (pi == null)
                    continue;

                if (pi.PropertyType.IsArray)
                {
                    string[] values = xmlAttribute.Value.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);

                    Type elementType = pi.PropertyType.GetElementType();
                    Array array = Array.CreateInstance(elementType, values.Length);
                    for (int i = 0; i < values.Length; i++)
                        array.SetValue(Utils.Parse(values[i], elementType), i);
                    pi.SetValue(obj, array);
                }
                else
                {
                    pi.SetValue(obj, Utils.Parse(xmlAttribute.Value, pi.PropertyType));
                }
            }

            foreach (XmlNode xmlNode in xmlMain.ChildNodes)
            {
                XmlElement xmlElement = xmlNode as XmlElement;
                if (xmlElement == null)
                    continue;

                PropertyInfo pi = type.GetProperty(xmlElement.Name);
                if (pi == null || !typeof(IList).IsAssignableFrom(pi.PropertyType))
                    continue;

                IList list = pi.GetValue(obj) as IList;
                list.Clear();
                Type[] genTypes = pi.PropertyType.GenericTypeArguments;
                if (genTypes != null && genTypes.Length > 0)
                {
                    foreach (XmlNode xmlNode2 in xmlElement.ChildNodes)
                    {
                        XmlElement xmlElement2 = xmlNode2 as XmlElement;
                        if (xmlElement2 == null)
                            continue;

                        if (genTypes[0] == typeof(string))
                            list.Add(xmlElement2.GetAttribute("Value"));
                        else
                        {
                            Object newObj = genTypes[0].GetConstructor(new Type[0]).Invoke(null);
                            XmlSerializer.Deserialize(xmlElement2, newObj);
                            list.Add(newObj);
                        }
                    }
                }
            }
        }

    }
}
